### PR TITLE
DAOS-14181 control: Remove bdev scan cache

### PR DESCRIPTION
**Pull request by tanabarr**
Instead of storing controller details in cache before starting engines,
reread info when scanning over dRPC. This enables details of
replacement controller to be returned after a hotplug event and
reduces some control-plane complexity.

Required-githooks: true

### Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [x] You are the appropriate gatekeeper to be landing the patch.
* [x] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [x] Githooks were used. If not, request that user install them and check copyright dates.
* [x] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [x] All builds have passed.  Check non-required builds for any new compiler warnings.
* [x] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [x] If applicable, the PR has addressed any potential version compatibility issues.
* [x] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [x] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/control/server/storage/provider_test.go
Comment (kjacque) [inline src/control/server/storage/provider_test.go:34]: Needs removal?

File: src/control/server/storage/bdev_test.go
Comment (kjacque) [inline src/control/server/storage/bdev_test.go:177]: Possible to break this up on word boundaries? Seems like it would make codespell happier, and it'd be a bit easier for humans to read.

File: src/control/server/instance_storage_test.go
Comment (kjacque) [inline src/control/server/instance_storage_test.go:2]: Revert? Looks like no code changes

File: src/control/server/instance_storage_rpc_test.go
Comment (kjacque) [inline src/control/server/instance_storage_rpc_test.go:105]: Test cases to be added?

File: src/control/server/instance_storage_rpc.go
Comment (kjacque) [inline src/control/server/instance_storage_rpc.go:232]: I know this is legal Go, but this multiple-assignment line is sort of confusing to read. Could it be broken up into separate assignments?

File: src/control/server/ctl_storage_rpc.go
Comment (kjacque) [inline src/control/server/ctl_storage_rpc.go:85]: Should we be returning the resp even if there's an error?

File: src/control/server/ctl_storage_rpc.go
Comment (kjacque) [inline src/control/server/ctl_storage_rpc.go:163]: What happens in this case? Does the drpc request return -DER_AGAIN?

Comment (kjacque) [reply src/control/server/ctl_storage_rpc.go:163]: I see potentially another race, where the code here sees no engines started, but then the engines start before the global bdev scan executes. What does that case look like to the user?

File: src/control/server/ctl_storage_rpc_test.go
Comment (kjacque) [inline src/control/server/ctl_storage_rpc_test.go:1528]: This wasn't meant to be removed, was it?

**Review by kjacque — CHANGES_REQUESTED** [review]

File: src/control/server/ctl_storage_rpc.go
Comment (tanabarr) [reply src/control/server/ctl_storage_rpc.go:85]: I don't really see the problem with doing so given we check for error before using a response in all the callers.

File: src/control/server/ctl_storage_rpc.go
Comment (tanabarr) [reply src/control/server/ctl_storage_rpc.go:163]: 1. In the situation where SPDK has claimed the controllers for the engine process but dRPC is not yet ready for comms, an "engine not ready" error should be returned. The window should be <<1s.
2. In the situation where this check results in a global bdev scan but by the time the scan executes the controllers have been claimed by the engine process, I think no results will be returned.

In terms of how to handle the above 2 scenarios I suggest the following:
- If "engine not ready" we issue a similar fault suggesting a retry.
- If no bdevs are found but some are expected (as bdev_list has been populated in config), fault suggesting retry is given.

I would prefer not to make it more complicated than this at the current stage to avoid the risk of pre-empting a solution for something that isn't a problem in practice.
Does the above suggestion sound acceptable for now?

File: src/control/server/ctl_storage_rpc.go
Comment (tanabarr) [reply src/control/server/ctl_storage_rpc.go:85]: done in the next commit

File: src/control/server/ctl_storage_rpc_test.go
Comment (tanabarr) [reply src/control/server/ctl_storage_rpc_test.go:1528]: nope, good catch, restored in the next commit, thanks

File: src/control/server/ctl_storage_rpc_test.go
Comment (tanabarr) [reply src/control/server/ctl_storage_rpc_test.go:1632]: nope, good catch, restored in the next commit, thanks

File: src/control/server/instance_storage_rpc.go
Comment (tanabarr) [reply src/control/server/instance_storage_rpc.go:232]: sure

File: src/control/server/instance_storage_test.go
Comment (tanabarr) [reply src/control/server/instance_storage_test.go:2]: done

File: src/control/server/storage/bdev_test.go
Comment (tanabarr) [reply src/control/server/storage/bdev_test.go:177]: tbh don't think that's an efficient use of time, would prefer not to unless you insist

File: src/control/server/storage/provider_test.go
Comment (tanabarr) [reply src/control/server/storage/provider_test.go:34]: done

File: src/control/server/ctl_storage_rpc.go
Comment (mjmac) [reply src/control/server/ctl_storage_rpc.go]: It is certainly best practice to check that the returned error was non-nil before using the response, but best practices aren't always followed. Better to return a nil response for errors so that any programming mistake is detected quickly during development (hopefully!).

File: src/control/server/storage/bdev_test.go
Comment (kjacque) [reply src/control/server/storage/bdev_test.go:176]: I just hate seeing red on the GHA. It's minor though, wouldn't block on it.

File: src/control/server/ctl_storage_rpc.go
Comment (kjacque) [reply src/control/server/ctl_storage_rpc.go:167]: Sounds good to me. I just wanted to make sure we weren't hiding an error or returning weird/confusing results.

Comment (mjmac) [reply src/control/server/ctl_storage_rpc.go:167]: I think that Kris raises a valid concern, particularly in the context of CI testing -- we don't want to introduce a new source of intermittent failures. Why should the retry error be exposed to the user vs. just handling the retry in code?

**Review by knard-intel — APPROVED** [review]
LGTM

File: src/mgmt/srv_query.c
Comment (kjacque) [inline src/mgmt/srv_query.c:773]: We should be avoiding strlen. Could we use strnlen with ADDR_STR_MAX_LEN?

File: src/control/server/instance_storage_rpc.go
Comment (kjacque) [inline src/control/server/instance_storage_rpc.go:263]: Need to remove? Could make it a trace level log if it's useful sometimes.

**Review by kjacque — APPROVED** [review]
Minor things, could be addressed in a follow on.

File: src/mgmt/srv_query.c
Comment (tanabarr) [reply src/mgmt/srv_query.c:773]: I will address when I have to re-push (after the parent lands).

File: src/control/server/instance_storage_rpc.go
Comment (tanabarr) [reply src/control/server/instance_storage_rpc.go:263]: I will address when I have to re-push (after the parent lands).

File: src/control/server/ctl_storage_rpc.go
Comment (tanabarr) [reply src/control/server/ctl_storage_rpc.go]: Changes to implement retries when scanning over provider and no results are returned and also to wait for drpc to be ready before scanning when engine is running have been applied in this PR.

**Comment by tanabarr** [comment]
@kjacque @knard-intel @mjmac I have rebased on master now the parent PR landed. The only changes since last review approvals are the 2 most recent commits. Other that the code is identical and conflicts have just been resolved.
Please review the most recent commits, the main change is resolving the VMD format failures which were as a result of removing the cache and not setting VMD-enabled flag. The solution involves passing the controller info explicitly and passing the enabled setting from harness storage provider to engine storage provider on engine creation.

**Review by kjacque — APPROVED** [review]
Overall LGTM. One minor comment, could be addressed in a follow-on.

File: src/control/server/instance_storage_rpc.go
Comment (tanabarr) [reply src/control/server/instance_storage_rpc.go:83]: I will update in https://github.com/daos-stack/daos/pull/13463 when I rebase.

**Comment by tanabarr** [comment]
CI failures all attributable to known issue https://daosio.atlassian.net/browse/DAOS-14718

**Review by mjmac — COMMENTED** [review]
The code looks fine overall, good to ultimately simplify the logic by avoiding the cache. I see that this was tested with `Features: control` ... Is that sufficient to have a high level of confidence that this won't introduce regressions on real hardware? Does that testing include VMD-enabled/disabled configurations and md-on-ssd?

**Comment by tanabarr** [comment]
> The code looks fine overall, good to ultimately simplify the logic by avoiding the cache. I see that this was tested with `Features: control` ... Is that sufficient to have a high level of confidence that this won't introduce regressions on real hardware? Does that testing include VMD-enabled/disabled configurations and md-on-ssd?


CI run https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-13385/20/pipeline/1200/ ran hardware lage, medium and medium+verbs. Medium runs with VMD AFAIK but we currently have MD-on-SSD disabled to reduce CI load. I will address your code comment and rerun with `Features: control` and `Test-nvme: auto_md_on_ssd`.

File: src/control/cmd/dmg/pretty/storage_nvme.go
Comment (tanabarr) [reply src/control/cmd/dmg/pretty/storage_nvme.go:297]: done

**Review by knard-intel — APPROVED** [review]

File: src/control/server/ctl_storage_rpc.go
Comment (kjacque) [inline src/control/server/ctl_storage_rpc.go:858]: Do we have reasons to be positive about this type underlying the interface? I know this is the type we actually have at runtime, but IMO we ought to check if the conversion worked, and return an error (or skip?) if not.

**Review by kjacque — COMMENTED** [review]
Only one comment, all other of my issues are resolved.

File: src/control/server/ctl_storage_rpc.go
Comment (tanabarr) [reply src/control/server/ctl_storage_rpc.go:858]: EngineInstance is the only type that satisfies the Engine interface I think this is okay

Comment (kjacque) [reply src/control/server/ctl_storage_rpc.go:858]: I do think the type ought to be checked anyway -- but I'm okay with it being addressed in a follow-on PR.

**Review by kjacque — APPROVED** [review]

**Comment by tanabarr** [comment]
@daos-stack/daos-gatekeeper I think this PR is ready for landing now thank you

File: src/control/server/ctl_storage_rpc.go
Comment (tanabarr) [reply src/control/server/ctl_storage_rpc.go:858]: PR-13463 is based on this PR branch and will need re-basing once this lands, I will add the check to that. Thanks Kris

**Review by mjmac — APPROVED** [review]

**Pull request merged**
